Name: Video Camera Flashlight
Version: 1.4-1
Author: Jason Robitaille
Description: Adds a flashlight toggle in the video recorder of the camera app  (MIT license)

--- .orig/usr/palm/applications/com.palm.app.camera/app/controllers/capture-assistant.js
+++ /usr/palm/applications/com.palm.app.camera/app/controllers/capture-assistant.js
@@ -14,6 +14,8 @@
 			FLASH_ON:   2
 		});		
 	
+		this.videaFlashState = CaptureAssistant.FLASH_OFF;
+		
 		this.photorollInfo = null;
 		
 		this._powerService = null;
@@ -233,6 +235,7 @@
 
 		this.controller.sceneElement.removeClassName('preview-off');
 		this.controller.get("camera-icon").style.display = "block";
+		this.loadVideoFlash();
 		llog("Controls are up");
 	},
 
@@ -909,6 +912,24 @@
 	 * @param {Event} event  The click event
 	 */
 	_handleFlashButton: function(event){
+		if(this.isVideoCapture) {
+			if(this.videaFlashState == CaptureAssistant.FLASH_OFF) {
+				this.videaFlashState = CaptureAssistant.FLASH_ON;
+				this.controller.get('flash-toggle').className = "flash-button flash-on";
+				this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+					method: 'setFlash',
+					parameters: {state:"on"}
+				});
+			} else {
+				this.videaFlashState = CaptureAssistant.FLASH_OFF;
+				this.controller.get('flash-toggle').className = "flash-button flash-off";
+				this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+					method: 'setFlash',
+					parameters: {state:"off"}
+				});
+			}
+			return;
+		}
 		if ( (this.flashState < CaptureAssistant.FLASH_OFF) || (this.flashState >= CaptureAssistant.FLASH_ON) ){
 			// Cycle order is off, auto, on.
 			this.flashState = CaptureAssistant.FLASH_OFF;
@@ -974,6 +995,26 @@
 		llog("Set Flash Mode to ", camera.flash);
 	},
 	
+	loadVideoFlash: function() {
+		if(this.isVideoCapture) {
+			if(this.videaFlashState == CaptureAssistant.FLASH_OFF) {
+				this.controller.get('flash-toggle').className = "flash-button flash-off";
+				this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+					method: 'setFlash',
+					parameters: {state:"off"}
+				});
+			} else {
+				this.controller.get('flash-toggle').className = "flash-button flash-on";
+				this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+					method: 'setFlash',
+					parameters: {state:"on"}
+				});
+			}
+		}
+	},
+	
+	
+	
 	/**
 	 * Handle the toggle event from the UI.
 	 * 
@@ -1004,7 +1045,7 @@
 		if (this.isVideoCapture) {
 			this.controller.sceneElement.removeClassName('camera');
 			this.controller.sceneElement.addClassName('camcorder');
-			
+			this.loadVideoFlash();
 		}
 		else {
 			this.controller.sceneElement.removeClassName('camcorder');
@@ -1007,6 +1048,18 @@
 			
 		}
 		else {
+			this.videaFlashState = CaptureAssistant.FLASH_OFF;
+			this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+				method: 'setFlash',
+				parameters: {state:"off"}
+			});
+			if (this.flashState == CaptureAssistant.FLASH_OFF) {
+				this.controller.get('flash-toggle').className = "flash-button flash-off";
+			} else if (this.flashState == CaptureAssistant.FLASH_ON) {
+				this.controller.get('flash-toggle').className = "flash-button flash-on";
+			} else if (this.flashState == CaptureAssistant.FLASH_AUTO) {
+				this.controller.get('flash-toggle').className = "flash-button flash-auto";
+			} 
 			this.controller.sceneElement.removeClassName('camcorder');
 			this.controller.sceneElement.addClassName('camera');
 		}
@@ -1074,6 +1127,10 @@
 	 */
 	_onBlur: function(event){
 		llog("Got a blur event");
+		this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+			method: 'setFlash',
+			parameters: {state:"off"}
+		});
 
 		if (this._blurHandlerTimeout){
 			llog("Re-initializing the blurHandler timeout");
--- .orig/usr/palm/applications/com.palm.app.camera/stylesheets/camera.css
+++ /usr/palm/applications/com.palm.app.camera/stylesheets/camera.css
@@ -129,6 +129,7 @@
 .camcorder.recording .capture-button.selected { background-position: 0px -400px; }
 
 .flash-button {
+    display: block !important;
     width: 50px;
     height: 50px;
     left: 50%;
@@ -141,8 +142,6 @@
     background: url(../images/flash-buttons.png) 0px 0px no-repeat;
 }
 
-.camcorder .flash-button { visibility: hidden; }
-
 .flash-button.selected { background-position: 0px -50px; }
 .flash-button.flash-on { background-position: 0px -100px; }
 .flash-button.flash-on.selected { background-position: 0px -150px; }
